python/ray/data/datasource/partitioning.py

```diff
@@ -0,0 +1,361 @@
+
+    Examples:
+        >>> # Serialize to JSON text.
+        >>> json.dumps(PartitionStyle.HIVE)  # "hive"
+
+        >>> # Deserialize from JSON text.
+        >>> PartitionStyle(json.loads('"hive"'))  # PartitionStyle.HIVE
+    """
+
+    HIVE = "hive"
+    DIRECTORY = "dir"
+
+
+@DeveloperAPI
+class PathPartitionBase:
+    """Base class for path-based partition formats.
+
+    Two path partition formats are currently supported - HIVE and DIRECTORY.
+
+    Path-based partition formats embed all partition keys and values directly in
+    their dataset file paths.
+    """
+
+    def __init__(
+        self,
+        style: PartitionStyle,
+        base_dir: Optional[str],
+        field_names: Optional[List[str]],
+    ):
+        """Creates a new path-based dataset partition scheme.
+
+        Args:
+            style: The partition style - may be either HIVE or DIRECTORY.
+            base_dir: "/"-delimited base directory that all partitioned paths should
+                exist under (exclusive).
+            field_names: The partition key field names (i.e. column names for tabular
+                datasets). Required when parsing DIRECTORY partitioned paths or
+                generating HIVE partitioned paths.
+        """
+        self._style = style
+        self._base_dir = base_dir or ""
+        self._normalized_base_dir = None
+        self._field_names = field_names
+
+    @property
+    def style(self) -> PartitionStyle:
+        return self._style
+
+    @property
+    def base_dir(self) -> str:
+        return self._base_dir
+
+    @property
+    def normalized_base_dir(self) -> Optional[str]:
+        return self._normalized_base_dir
+
+    @property
+    def field_names(self) -> Optional[List[str]]:
+        return self._field_names
+
+    def _normalize_base_dir(self, filesystem: "pyarrow.fs.FileSystem"):
+        """Normalizes the partition base directory for compatibility with the
+        given filesystem.
+
+        This should be called once a filesystem has been resolved to ensure that this
+        base directory is correctly discovered at the root of all partitioned file
+        paths.
+        """
+        from ray.data.datasource.file_based_datasource import (
+            _resolve_paths_and_filesystem,
+        )
+
+        paths, _ = _resolve_paths_and_filesystem(
+            self._base_dir,
+            filesystem,
+        )
+        assert (
+            len(paths) == 1
+        ), f"Expected 1 normalized base directory, but found {len(paths)}"
+        normalized_base_dir = paths[0]
+        if len(normalized_base_dir) and not normalized_base_dir.endswith("/"):
+            normalized_base_dir += "/"
+        self._normalized_base_dir = normalized_base_dir
+
+
+@DeveloperAPI
+class PathPartitionGenerator(PathPartitionBase):
+    """Callable that generates directory paths for path-based partition formats.
+
+    Path-based partition formats embed all partition keys and values directly in
```

Comment:
I wonder what are the non-path based partition look like? Relying on metadata to identify what are the partitions?

Reply:
Metadata files co-located with the data or maintained in dedicated external storage (e.g. distributed KV stores) both seem to be common options. For example, Apache Iceberg stores per-file partition values in [manifest files](https://iceberg.apache.org/spec/#manifests). Meanwhile, Delta Lake [stores partitioned files under Hive-like paths by default](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#data-files), but relies on [transaction log checkpoint files](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#checkpoint-schema) as the authoritative source for reading actual partition values. Meanwhile, Amazon's data catalog stores partition value metadata in external DynamoDB tables.

IMO, the common characteristic between all of these (including path-based partitioning) is that they all support the concept of retrieving or associate partition keys and values with individual dataset files, while the main differences include the logic to map partition KVs to/from files, and the type used to losslessly read/write partition values (where path-based partitioning relies on an undefined/external spec for serializing/deserializing partition values to/from strings).